sympy/matrices/matrices.py

```diff
@@ -783,6 +784,26 @@ def __repr__(self):
         return "MatrixKind(%s)" % self.element_kind
 
 
+@Mul._kind_dispatcher.register(MatrixKind, _NumberKind)
+def mat_num_mul(k1, k2):
+    """Return MatrixKind. The element kind is selected by recursive dispatching."""
+    # Deal with Mul._kind_dispatcher's commutativity
+    elemk = Mul._kind_dispatcher(k1.element_kind, NumberKind)
+    return MatrixKind(elemk)
+
+@Mul._kind_dispatcher.register(_NumberKind, MatrixKind)
+def num_mat_mul(k1, k2):
```

Comment:
I see `num_mat_mul` and `num_mul_mat` both encode the fact that `number*matrix ==> matrix`. Is there a way to avoid defining this rule twice? Ideally you should be able to tell SymPy that `number*matrix == matrix*number == matrix` only once.

Reply:
If we have `N` kinds, the current approach will require to define `N**2` doubly-dispatched methods.

Have you had a look at [promote_rule](https://docs.julialang.org/en/v1/manual/conversion-and-promotion/#Defining-Promotion-Rules) in Julia? They still do double-dispatching, but not on every single of the `N**2` combinations of types. The define which types can be converted into other ones and what type they should be promoted to.

How can this be adapted to kinds?

Kind promotion could be defined as `kind_promote_rule(target_kind, starting_kind)`. For example
- `kind_promote_rule(MatrixKind, NumberKind) ==> MatrixKind`
- `kind_promote_rule(NumberKind, MatrixKind) ==> not defined`

At this point, whenever you have `Mul[Kind1, Kind2]` you would first try to promote `Kind1` to `Kind2`, if this fails you can later try to promote `Kind2` to `Kind1`. If both fail, the two objects have incompatible kinds and cannot be multiplied together. OTOH, this method only requires to define one promotion rule (i.e. _NumberKind ==> MatrixKind_, instead of having to define the dispatcher twice).

I'm insisting on this because I fear we will need an enormous amount of dispatched methods if we require to define all `N**2` pairs.

Reply:
Please see [this review](https://github.com/sympy/sympy/pull/20659#discussion_r550651935).

Reply:
I will see to it.

Reply:
There won't be that many combinations of kinds where mixing in a Mul actually makes sense.

At least in the way that I've been thinking about Kinds the idea is not to have many of them in total anyway. For example it might make sense to have a Kind for a quantity that has both a magnitude and a unit like `6*metres`. I wouldn't propose to have a different Kind for every different unit though: there would be just one Kind like QuantityKind to show that it is different from NumberKind and to make it clear that it has a unit in the same way that a Matrix has a shape. Then we can see that QuantityKind is an invalid argument for a transcendental function like `exp`.

What examples of Kinds that go together in a Mul can you think of? Probably any Expr Kind can be multiplied by at least some examples from NumberKind but otherwise what combinations of non-NumberKind can there reasonably be?

QuantityKind leads to an interesting question though:
```
In [12]: x = Symbol('x')

In [13]: from sympy.physics.units import m

In [14]: (x*m + 4*m)/(3*m)
Out[14]: 
x⋅meter + 4⋅meter
─────────────────
     3⋅meter

In [15]: expand(_)
Out[15]: 
x   4
─ + ─
3   3
```
Really the `Kind` here is `NumberKind` although it's a `Mul` of `QuantityKind`. Maybe there would have to be a `QuantityMul` that could define `.kind` properly (as well as the unit of the Mul).

We can change the implementation of `.kind` later if it turns out to be cumbersome. For now I think it would just be good to get the basic idea in so that we can use it and see what problems are encountered along the way.

Reply:
> Maybe there would have to be a `QuantityMul` that could define `.kind` properly (as well as the unit of the Mul).

Wasn't _kind_ introduced in order to get rid of subclasses of `Mul`?

Reply:
> There won't be that many combinations of kinds where mixing in a Mul actually makes sense.

Even if you have only 4 or 5 kinds, you would need to double-dispatch 4-5 multimethods instead of 16-25 ones.

Also think of the complications of adding new kinds in the future. If you have 4 kinds, adding a new kind would require the developer to add 9 multimethods (all compatible pairs), instead of just a few ones.

Furthermore, I don't see the point of redefining the same method multiple times, just because both pairs `(Kind1, Kind2)` and `(Kind2, Kind1)` have to be dispatched.

Reply:
I implemented this in 6564d16.

Reply:
By the way, I just realized that @asmeurer opened [this issue](https://github.com/mrocklin/multipledispatch/issues/110).

Reply:
> Wasn't _kind_ introduced in order to get rid of subclasses of `Mul`?

There could still be a reason for having subclasses in some cases but the Kind system should mean that we can make most things work without being needing to create subclasses (for everything, not just Mul). One thing `Kind` does is provides a way to replace the post-processors with something more systematic for a `Kind` that does have an associated `Mul` class (the `kind` can be used to dispatch to `MatMul`). It can also make `Mul` itself with different kinds more manageable.

> Even if you have only 4 or 5 kinds, you would need to double-dispatch 4-5 multimethods instead of 16-25 ones.

The question is not how many kinds there are but how many combinations of them are actually meaningful in a product. The mechanism here makes it possible to define meaningful combinations if you want but in practice there aren't really many meaningful combinations. The only examples we have so far are MatrixKind and NumberKind. The other kinds like BooleanKind, SetKind etc can't go in a Mul at all.

Reply:
I think *some* subclasses still need to be defined, though. For example non-argument `Add()` returns `0` but we expect addition between (2,2) matrices with no argument to return `Matrix([[0,0], [0,0]])`.

It is true that most combination of `Kinds` do not need to be defined. However having just one less handler (`num_mat_mul & mat_num_mul` -> only `num_mat_mul`) still seems good so I implemented this. I will mark this as resolved.

Reply:
> I think _some_ subclasses still need to be defined, though. For example non-argument `Add()` returns `0` but we expect addition between (2,2) matrices with no argument to return `Matrix([[0,0], [0,0]])`, which is `MatAdd()`. Having only `Add` will make it hard to do such thing easily.

That's actually an argument to include the essential information into `MatrixKind`. For example, `MatrixKind[NumberKind](2, 2)` could be the kind of matrices of numbers with shape `(2, 2)`. At this point `Add.kind` should be able to return the correct shape.

IIRC, you implemented `.kind` as a singleton. The approach I've just suggested would probably require removing the singleton pattern from `.kind`.

Reply:
I don't think that we need a separate kind for each shape of matrix but there could be a way to use the kind object to do this. The `Add` function could call `MatrixKind[NumberKind]._eval_add(args)` or something like that. The shape would have to be inferred from the `args` and if they all cancel then an appropriately shaped zero matrix can be returned.

Reply:
Oh, okay. Perhaps we can remove subclasses that way.

Reply:
I actually think that it makes sense to have the shape info in MatrixKind. In this way, when you add two matrices, you can check immediately if the kinds are compatible without recursing to the elements.

Consider Add(Add(...), Add(...)) as an example.

Having shape info allows you to tell immediately if ArrayKind and MatrixKind may be summed it not.

Reply:
If you found a way to remove subclasses, that would be wonderful.

Reply:
> IIRC, you implemented .kind as a singleton. The approach I've just suggested would probably require removing the singleton pattern from .kind.

`Kind` returns same object for same arguments. Incorporating shape into `MatrixKind` can be like `MatrixKind(NumberKind, (2, 2))`, which preserves this behavior.

Reply:
I think that would be overcomplicating things. In some cases the shape can be symbolic and then inferring the compatibility of the kinds becomes undecidable.
```python
In [26]: a, b, c, d = symbols('a, b, c, d')

In [27]: M = MatrixSymbol('M', a, b)

In [28]: N = MatrixSymbol('N', c, d)

In [29]: Eq(M, N)
Out[29]: False

In [30]: Eq((a, b), (c, d))
Out[30]: (a, b) = (c, d)
```
Unfortunately there is a bug which gives False for `Eq(M, N)` which should really remain unevaluated (symbolic shapes are not currently well handled).

It should be possible though to say that if `M.kind is not N.kind` then `Eq(M, N) is S.false`. That won't be possible if undecidable symbolic determination is involved in comparing kinds.

Reply:
> It should be possible though to say that if `M.kind is not N.kind` then `Eq(M, N) is S.false`.

Well, what about `Eq(x, 1)`? It does not sound so obvious to me. Objects can still be equal even if of different kind.

Reply:
Currently, `Symbol` with `commutative=True` is generally used as undecided numeric scalar so we designed it to have `NumberKind`. Thus we don't have to care about `Eq(x, 1)`.

Too long has SymPy used mathematically ambiguous object to represent *anything*. We should narrow down the usage of each objects by now, and `Kind` design will be the first step to it.

By the way, I think the discussion here is completely unrelated to `KindDispatcher`. @Upabjojr, do you approve this PR? If you do, then we can get this merged first and continue the discussion.

Reply:
It should be possible to create e.g. symbols like `Symbol('x', kind=BooleanKind)` and there should be perhaps a `BooleanSymbol` class and a `boolean_symbols` function for this. A symbol will need to have an associated kind. Default Symbol will be NumberKind. That is one of the implications of the kind system as implemented so far.

Reply:
> @Upabjojr, do you approve this PR?

Yes, I do.

Reply:
Thanks! I will get rid of postprocessors first with a few more PR, then I'll attend to this subclass topic. Hopefully we can have more insight with working examples.

Reply:
Okay, great. Let's get this merged and take this forward.

Reply:
Okay great. Let's get this merged then.